packages/react-dom/src/__tests__/ReactDOMServerIntegrationHooks-test.internal.js

```diff
@@ -843,4 +853,384 @@ describe('ReactDOMServerHooks', () => {
+        const idOne = useOpaqueReference();
+        const idTwo = useOpaqueReference();
+        return (
+          <div>
+            <div aria-labelledby={idOne} />
+            <div id={idOne} />
+            <span aria-labelledby={idTwo} />
+            <span id={idTwo} />
+          </div>
+        );
+      }
+
+      try {
+        await clientRenderOnBadMarkup(<App />);
+      } catch (x) {
+        // We expect this to trigger the BadMarkupExpected rejection.
+        if (!(x instanceof BadMarkupExpected)) {
+          // If not, rethrow.
+          throw x;
+        }
+      }
+    });
+
+    it('useOpaqueReference does not change id even if the component updates during client render', async () => {
+      let _setShowId;
+      function App() {
+        const id = useOpaqueReference();
+        const [showId, setShowId] = useState(false);
+        _setShowId = setShowId;
+        return (
+          <div>
+            <div aria-labelledby={id} />
+            {showId && <div id={id} />}
+          </div>
+        );
+      }
+
+      const domNode = await clientCleanRender(<App />);
+
+      expect(domNode).toMatchInlineSnapshot(`
+        <div>
+          <div
+            aria-labelledby="c_0"
+          />
+        </div>
+    `);
+
+      await ReactTestUtils.act(async () => _setShowId(true));
+
+      expect(domNode).toMatchInlineSnapshot(`
+          <div>
+            <div
+              aria-labelledby="c_0"
+            />
+            <div
+              id="c_0"
+            />
+          </div>
+      `);
+    });
+
+    it('useOpaqueReference: IDs match when, after hydration, a new component that uses the ID is rendered', async () => {
+      let _setShowDiv;
+      function App() {
+        const id = useOpaqueReference();
+        const [showDiv, setShowDiv] = useState(false);
+        _setShowDiv = setShowDiv;
+
+        return (
+          <div>
+            <div id={id}>Child One</div>
+            {showDiv && <div id={id}>Child Two</div>}
+          </div>
+        );
+      }
+
+      const container = document.createElement('div');
+      document.body.append(container);
+
+      container.innerHTML = ReactDOMServer.renderToString(<App />);
+      ReactDOM.hydrate(<App />, container);
+      expect(container).toMatchInlineSnapshot(`
+        <div>
+          <div
+            data-reactroot=""
+          >
+            <div
+              id="s_0"
+            >
+              Child One
+            </div>
+          </div>
+        </div>
+      `);
+      await ReactTestUtils.act(async () => {
+        _setShowDiv(true);
+      });
+      expect(container).toMatchInlineSnapshot(`
+        <div>
+          <div
+            data-reactroot=""
+          >
+            <div
+              id="c_0"
+            >
+              Child One
+            </div>
+            <div
+              id="c_0"
+            >
+              Child Two
+            </div>
+          </div>
+        </div>
+      `);
+    });
```

Comment:
We should add a similar test where Child One is in a Suspense boundary that is still suspended. I.e. if we're only partially hydrated. Then we do an update that shows the second Child. This should suspend and wait for the suspense boundary to resolve before letting the update through.

It should not leave them inconsistent.

That test would ensure that we don't forget about this case and try to implement the "patching strategy approach" in the future.

https://github.com/facebook/react/pull/17322/files#r354445736